feat(batch): new BatchProcessor for SQS, DynamoDB, Kinesis

heitorlessa opened the pull request:
**Issue #, if available:** #858

## Description of changes:

Implements new Batch Processor following the [new built-in partial response handling](https://aws.amazon.com/about-aws/whats-new/2021/11/aws-lambda-partial-batch-response-sqs-event-source/) for Lambda SQS, DynamoDB Stream, and Kinesis Data Streams. 

**Checklist**

<!--- Leave unchecked if your change doesn't seem to apply --> 

* [x] [Meet tenets criteria](https://awslabs.github.io/aws-lambda-powertools-python/#tenets)
* [x] Update tests
* [x] Update docs
* [x] PR title follows [conventional commit semantics](https://github.com/awslabs/aws-lambda-powertools-python/blob/376ec0a2ac0d2a40e0af5717bef42ff84ca0d1b9/.github/semantic.yml#L2)
* [x] Add DynamoDB Stream support
* [x] Add Kinesis Data Stream support
* [x] Think of a better name other than `.report()` -> `response` to match Java counterpart
* [x] Support event source data classes by default
* [x] Mypy support
* [x] Review with @cakepietoast whether we still need `suppress_exception` behaviour
* [x] Bring-your-own-model Pydantic Parser as an option
* [x] Test for incorrect data/payload that won't be parsed (either Model or Data Class)
* [x] Raise `BatchProcessingError` when all batch records fail to ensure failure arises

**Tasks to be created in separate PR to ease reviewing**

* [ ] Support for handling permanent exceptions (might be for the release after next)

### UX

**As a decorator**

```python
from aws_lambda_powertools.utilities.batch import batch_processor, BatchProcessor, EventType
from aws_lambda_powertools.utilities.data_classes.sqs_event import SQSRecord

processor = BatchProcessor(event_type=EventType.SQS) # or, EventType.KinesisDataStreams, EventType.DynamoDB

def record_handler(record: SQSRecord): # or DynamoDBRecord, KinesisStreamRecord
    return "success"

@batch_processor(record_handler=record_handler, processor=processor)
def lambda_handler(event, context):
    return processor.response() # this is necessary to ensure Lambda returns the correct response
```

**As a context manager**, in case you need full access to the batch processed, or handle exceptions yourself

```python
from aws_lambda_powertools.utilities.batch import batch_processor, BatchProcessor, EventType

processor = BatchProcessor(event_type=EventType.SQS) 


def lambda_handler(event, context):
    records = event["Records"]
    with processor(records, record_handler) as batch:
        processed_messages = batch.process()
    
    return batch.response() # this is necessary to ensure Lambda returns the correct response
```

## Breaking change checklist

<!--- Ignore if it's not a breaking change -->

**RFC issue #**:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

heitorlessa commented:
@cakepietoast ready to review implementation before it gets bigger with mypy and doc stuff.

ran-isenberg commented:
@heitorlessa  as discussed, here's the sqs example:

@validator('body', pre=True)
    def transform_body_to_dict(cls, value):
        return json.loads(value)

gmcrocetti commented on aws_lambda_powertools/utilities/batch/base.py:
```diff
@@ -146,3 +184,108 @@ def batch_processor(
+        """
+        self.event_type = event_type
+        self.model = model
+        self.batch_response = self.DEFAULT_RESPONSE
```
Hey @heitorlessa , don't know if this class is meant to be extended. Case it's not, following comment won't make much sense but anyways:
Do you think it's worth being more defensive in here by adding a deepcopy ?
```suggestion
        self.batch_response = deepcopy(self.DEFAULT_RESPONSE)
```
Just to avoid a scenario where someone extends from this class and mess with the `DEFAULT_RESPONSE` through `self.batch_response` (`self.batch_response.update({"batchItemFailures": []})` for instance ) - or any sort of changes...

cakepietoast commented on aws_lambda_powertools/utilities/batch/base.py:
```diff
@@ -72,7 +110,7 @@ def __call__(self, records: List[Any], handler: Callable):
         return self
 
-    def success_handler(self, record: Any, result: Any):
+    def success_handler(self, record, result: Any) -> SuccessCallback:
```
Would `SuccessResponse` be a more accurate name?

heitorlessa replied:
YES! Good catch -- will change that

heitorlessa commented:
🤦‍♂️ yes! Not sure why I used Callback

On Wed, 15 Dec 2021 at 09:54, Tom McCarthy ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In aws_lambda_powertools/utilities/batch/base.py
> <https://github.com/awslabs/aws-lambda-powertools-python/pull/886#discussion_r769395628>
> :
>
> > @@ -72,7 +110,7 @@ def __call__(self, records: List[Any], handler: Callable):
>          self.handler = handler
>          return self
>
> -    def success_handler(self, record: Any, result: Any):
> +    def success_handler(self, record, result: Any) -> SuccessCallback:
>
> Would SuccessResponse be a more accurate name?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/awslabs/aws-lambda-powertools-python/pull/886#pullrequestreview-832497325>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZPQBFPYJNIEG5KWW7JBGDURBJTJANCNFSM5JZVI3AA>
> .
>

heitorlessa commented:
Great point. It’s good to anticipate someone will subclass it sooner or
later

On Tue, 14 Dec 2021 at 15:54, Guilherme Martins Crocetti <
***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In aws_lambda_powertools/utilities/batch/base.py
> <https://github.com/awslabs/aws-lambda-powertools-python/pull/886#discussion_r768745433>
> :
>
> > +class BatchProcessor(BasePartialProcessor):
>
> +    DEFAULT_RESPONSE: Dict[str, List[Optional[dict]]] = {"batchItemFailures": []}
>
> +
>
> +    def __init__(self, event_type: EventType, model: Optional["BatchTypeModels"] = None):
>
> +        """Process batch and partially report failed items
>
> +
>
> +        Parameters
>
> +        ----------
>
> +        event_type: EventType
>
> +            Whether this is a SQS, DynamoDB Streams, or Kinesis Data Stream event
>
> +        model: Optional["BatchTypeModels"]
>
> +            Parser's data model using either SqsRecordModel, DynamoDBStreamRecordModel, KinesisDataStreamRecord
>
> +        """
>
> +        self.event_type = event_type
>
> +        self.model = model
>
> +        self.batch_response = self.DEFAULT_RESPONSE
>
>
> Hey @heitorlessa <https://github.com/heitorlessa> , don't know if this
> class is meant to be extended. Case it's not, following comment won't make
> much sense but anyways:
> Do you think it's worth being more defensive in here by adding a deepcopy ?
> ⬇️ Suggested change
>
> -        self.batch_response = self.DEFAULT_RESPONSE
>
> +        self.batch_response = deepcopy(self.DEFAULT_RESPONSE)
>
>
> Just to avoid a scenario where someone extends from this class and mess
> with the DEFAULT_RESPONSE through self.batch_response (self.batch_response.update({"batchItemFailures":
> []}) for instance ) - or any sort of changes...
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/awslabs/aws-lambda-powertools-python/pull/886#pullrequestreview-831612986>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZPQBDWTSDMHXO55IOJODDUQ5LBTANCNFSM5JZVI3AA>
> .
>

jplock commented on aws_lambda_powertools/utilities/batch/base.py:
```diff
@@ -146,3 +185,120 @@ def batch_processor(
     return handler(event, context)
+
+
+class BatchProcessor(BasePartialProcessor):
```
would it make sense to have separate processors for each event type (SQS, DynamoDB or Kinesis) instead of growing the complexity of this class?  Then you could encapsulate the failure collection in the specific processor.

heitorlessa replied:
That was the initial version we wanted to implemented - a KinesisDataStreamProcessor, DynamoDB... then @cakepietoast argued that this was gonna confuse customers with other available processors (Sqs, PartialProcessor, BaseProcessor), as we can only deprecate them in v2.

I'm 50/50 here if I'm honest. I prefer a separate one but I also can see customers easily confused of which one to pick despite docs change I'm gonna make.

Implementation wise, this will remain stable. The only two changes I can anticipate is 1/ supporting the new Permanent Exception parameter, and 2/ raising a descriptive exception in case we reach an AttributeError when collecting message id/sequence number from a malformed event/model.

jplock replied:
What if we changed the nomenclature to be “producer” and “consumer” (these processors would be consumers). I had that other idea earlier to make it easier to use the SQS and DynamoDB batch write methods taking into account their batch sizes, those could be “producers” 🤷‍♂️

heitorlessa replied:
Where would the producer sit and what would be its responsibilities?

For that suggestion on partitioning, we **should** add it to the Event Source Data Class as it's a no brainier.

I **think** the word Consumer wouldn't be explicit enough on the capabilities Batch provide - maybe something else then?

Features

- Transform incoming records into Event Source Data Class or Pydantic Models
- Call an user defined function for each record in the batch
- Keep track of exceptions raised for any record
- For partial successes, extract message IDs (SQS) or sequence numbers (DynamoDB & Kinesis) and build a response to the new BatchItemIdentifiers contract
- Future: If a raised exception  matches the permanent_exceptions set, send these records to the configured DLQ in batches of 10

heitorlessa commented:
Docs finally done - @michaelbrewer @ran-isenberg could you have a quick pass before we release it tomorrow? All I need to do now is merge and start writing the release notes.

Here's what I've changed:

- [x] Updated the docs to reflect the new `BatchProcessor` in favour of `sqs_batch_processor` and `PartialSQSProcessor`
- [x] Added a Migration guide for those using `sqs_batch_processor` and `PartialSQSProcessor`
- [x] Added Caveats section in case people use Tracer response auto-capture feature with this utility in large batch sizes
- [x] Added sample event and sample response for each
- [x] Added SAM infrastructure for SQS, Kinesis and DynamoDB along with DLQ and IAM Permissions
- [x] Added an Extensibility section
- [x] Added Pydantic section
- [x] Updated Create your own processor
- [x] Updated Key features
- [x] Updated Intro
- [x] Recreated Background
- [x] Repurposed a few previous sections like Sentry integration under Legacy

michaelbrewer commented:
@heitorlessa  super cool. i will try to play around with it once merged

ran-isenberg commented:
@heitorlessa  looks great, thank you very much for putting the effort supporting both parser and data classes.
